man/mdio.8
.Dd September 17, 2021
.Dt MDIO 8 SMM
.Os Linux
.Sh NAME
.Nm mdio
.Nd Direct MDIO bus access
.Sh SYNOPSIS
.Nm mdio
.Op Ar bus Op Ar device Op Ar operation
.Nm mdio
.Op Fl h | Fl v
.Sh DESCRIPTION
Without any arguments, all available MDIO buses are listed. Supplying only
.Ar bus
(see
.Sx Buses )
will list any Clause 22 addressable devices on it. Omitting
.Ar operation
prints a brief status message for the
.Ar device
(see
.Sx Devices )
if the driver supports it. Otherwise, perform
.Ar operation
(see
.Sx Operations )
on
.Ar device
attached to
.Ar bus
.Ss Options
.Bl -tag
.It Fl h
Print usage message and exit.
.It Fl v
Print version information and exit.
.El
.Ss Buses
Denoted by their sysfs names as listed in
.Pa /sys/class/mdio_bus/ .
As these are typically very awkward to type,
.Xr glob 3
patterns may be used to abbreviate them (see
.Sx EXAMPLES )
.Ss Devices
Multiple types of devices are supported via pluggable
drivers. Different devices will use different addressing schemes. If
you are unsure about which type of device you have, using the
.Cm phy
driver is probably a good bet if your PHY runs at speeds \(<= 1 Gbps. If your PHY is capable of speeds > 1
Gbps, you will most likely want to use the
.Cm mmd
driver.
.Pp
The following devices are supported:
.Bl -tag -offset 2n
.It Cm phy Ar PHYAD
Clause 22 (MDIO) PHY using address
.Ar PHYAD .
.Pp
.Bl -tag -compact
.It Ar PHYAD
:= 0-31
.It Ar REG
:= 0-31
.El
.It Cm mmd Ar PRTAD:DEVAD
Clause 45 (XMDIO) MMD using address
.Ar PRTAD:DEVAD .
.Pp
.Bl -tag -compact
.It Ar PRTAD
:= 0-31
.It Ar DEVAD
:= 0-31
.It Ar REG
:= 0-0xffff
.El
.It Cm mmd-c22 Ar PRTAD:DEVAD
Clause 45 (XMDIO) MMD, accessed over Clause 22 registers 13 and 14,
using address
.Ar PRTAD:DEVAD .
.Pp
.Bl -tag -compact
.It Ar PRTAD
:= 0-31
.It Ar DEVAD
:= 0-31
.It Ar REG
:= 0-0xffff
.El
.It Cm mva Ar PHYAD
Marvell Alaska (mv88e1xxx) paged PHY. Register 22 is a paging register
used to expand the register space.
.Pp
.Bl -tag -compact
.It Ar PHYAD
:= 0-31
.It Ar REG
:= PAGE:0-31
.It Ar PAGE
:= 0-255|copper|fiber
.El
.It Cm mvls Ar ID
Marvell LinkStreet (mv88e6xxx) Ethernet switch. If ID = 0, single-chip
addressing is used; if ID \(!= 0, multi-chip addressing is used.
.Pp
.Bl -tag -compact
.It Ar ID
:= 0-31
.It Ar REG
:= PORT:0-31
.It Ar PORT
:= 0-31|g1|g2
.El
.It Cm xrs Ar PHYAD
Arrow/Flexibilis XRS700x Ethernet switch. Register space has a stride
of 2, i.e. only even numbered registers are valid.
.Pp
.Bl -tag -compact
.It Ar PHYAD
:= 0-31
.It Ar REG
:= 0-0xffffffff
.El
.El
.Ss Operations
The following operations are supported:
.Bl -tag -offset 2n
.It Cm raw Ar REG Op Ar DATA[/MASK]
Raw register access. Without
.Ar DATA ,
.Ar REG
is read.
.Ar DATA
without
.Ar MASK
will perform a single write to
.Ar REG .
.Ar DATA
with
.Ar MASK
will perform a read/mask/write
sequence.
.Pp
.Bl -tag -compact
.It Ar REG
Determined by the device type (see
.Sx Devices )
.It Ar DATA
:= 0-0xffff
.It Ar MASK
:= 0-0xffff
.El
.It Cm bench Ar REG Op Ar DATA
Benchmark read performance.
If
.Ar DATA
is supplied, it is written to
.Ar REG ,
otherwise the current value in
.Ar REG
is read.
.Ar REG
is then read 1000 times. Any unexpected values are reported, along
with the total time.
.Pp
.Bl -tag -compact
.It Ar REG
Determined by the device type (see
.Sx Devices )
.It Ar DATA
:= 0-0xffff
.El
.El
.Sh EXAMPLES
.Pp
Show all available buses:
.Bd -literal -offset 2n
~# mdio
30be0000.ethernet-1
fixed-0
.Ed
.Pp
List all Clause 22 addressable devices on a bus (using
.Xr glob 3
pattern to abbreviate bus name):
.Bd -literal -offset 2n
~# mdio 3*
 DEV      PHY-ID  LINK
0x01  0x01410dd0  up
.Ed
.Pp
Read register 2 from PHY 1:
.Bd -literal -offset 2n
~# mdio 3* phy 1 raw 2
0x0141
.Ed
.Pp
Perform a reset on PHY 1:
.Bd -literal -offset 2n
~# mdio 3* phy 1 raw 0 0x8000/0x7fff
.Ed
.Pp
Read register 0x1000 from MMD 4 on PHY 9:
.Bd -literal -offset 2n
~# mdio 3* mmd 9:4 raw 0x1000
0x2040
.Ed
.Pp
Read status register from the copper page of an Alaska PHY:
.Bd -literal -offset 2n
~# mdio 3* mva 1 raw copper:1
0x796d
.Ed
.Pp
Set the device number, of LinkStreet switch 4, to 10:
.Bd -literal -offset 2n
~# mdio 3* mvls 4 raw g1:28 0xa/0xfff0
.Ed
.Sh SEE ALSO
.Xr mvls 8
.Xr mdio-netlink 9
.Sh STANDARDS
IEEE Std 802.3-2018
.Bl -bullet -compact
.It
Clause 22 (MDIO)
.It
Clause 45 (XMDIO)
.El
.Sh AUTHORS
.An Tobias Waldekranz Aq Mt tobias@waldekranz.com
.Sh CAVEATS
In addition to the userspace parts of
.Nm ,
the
.Xr mdio-netlink 9
kernel module must also be installed.



ChangeLog.md
ChangeLog
=========

All notable changes to the project are documented in this file.

[v1.1.1] - 2022-05-23
---------------------

Tiny bugfix release.

### Fixed
- mdio: The bench operation is now much more reliable when stacked on
  other devices than regular PHYs (e.g. paged PHYs or Marvell
  switches).
- mvls: The STU can now be dumped chips from the Peridot generation.

[v1.1.0] - 2022-05-04
---------------------

A sprawling release, adding various mvls related introspection
features. mvls also gains a JSON output format.

### Added
- mvls: The STU can now be dumped (requires Linux 5.17 or later). This
  is useful now that mv88e6xxx supports offloading of MST states
- mvls: Output can now be formatted as JSON for easier scripting
- mdio: mvls: A subset of MIB counters can now be dumped. This let's
  you get at counters for DSA ports, which are not reachable from
  ethtool
- mdio: mvls: The LAG mask and LAG map tables can now be dumped
- mdio: Improve usage message by including the examples from the
  manual

[v1.0.1] - 2021-11-26
---------------------

Primarily fixes a few issues in the kernel module that were found
during a quick review from Russell King:

https://lore.kernel.org/netdev/YYPThd7aX+TBWslz@shell.armlinux.org.uk/
https://lore.kernel.org/netdev/YYPU1gOvUPa00JWg@shell.armlinux.org.uk/

### Added
- mdio: The mvls subcommand now supports flushing the ATU

### Fixed
- mdio-netlink: Plug some glaring holes around integer overflows of
  the PC.
- mdio-netlink: Release reference to MDIO bus after a transaction
  completes.


[v1.0.0] - 2021-09-17
---------------------

### Added
- Basic usage text, `mvls -h`
- Manuals

### Changes
- Reworked command syntax to be more ergonomic
- Improved error output

### Fixes
- Fix #4: buffer alignment on Arm9 systems

### Removed
- References to the dump operation in mdio(8) which is not supported
  at the moment

v1.0.0-beta1 - 2021-05-21
-------------------------

Initial public release.


[UNRELEASED]: https://github.com/wkz/mdio-tools/compare/1.0.0-beta1...HEAD
[v1.0.0]:     https://github.com/wkz/mdio-tools/compare/1.0.0-beta1...1.0.0


src/mdio/mdio.c
#include <assert.h>
#include <errno.h>
#include <glob.h>
#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <libmnl/libmnl.h>
#include <linux/genetlink.h>
#include <linux/mdio.h>
#include <linux/mdio-netlink.h>
#include <sys/types.h>
#include <sys/wait.h>

#include "mdio.h"

static int mdio_parse_bus_cb(const char *bus, void *_id)
{
	char **id = _id;

	*id = strdup(bus);
	return 1;
}

int mdio_parse_bus(const char *str, char **bus)
{
	*bus = NULL;
	mdio_for_each(str, mdio_parse_bus_cb, bus);

	if (*bus)
		return 0;

	fprintf(stderr, "ERROR: \"%s\" does not match any known MDIO bus", str);
	return ENODEV;
}

int mdio_parse_dev(const char *str, uint16_t *dev, bool allow_c45)
{
	unsigned long p, d = 0;
	char *end;

	p = strtoul(str, &end, 0);
	if (!end[0]) {
		allow_c45 = false;
		goto c22;
	}

	if (end[0] != ':')
		/* not clause 45 either */
		goto err_invalid;

	if (!allow_c45) {
		fprintf(stderr, "ERROR: Clause-45 addressing not allowed\n");
		return EINVAL;
	}

	d = strtoul(end + 1, &end, 0);
	if (end[0])
		goto err_invalid;

	if (allow_c45 && (d > 31)) {
		fprintf(stderr, "ERROR: Device %lu is out of range [0-31]\n", d);
		return ERANGE;
	}

c22:
	if (p > 31) {
		fprintf(stderr, "ERROR: %s %lu is out of range [0-31]\n",
			allow_c45 ? "Port" : "Device", d);
		return ERANGE;
	}

	*dev = allow_c45 ? mdio_phy_id_c45(p, d) : p;
	return 0;

err_invalid:
	fprintf(stderr, "ERROR: \"%s\" is not a valid device\n", str);
	return EINVAL;
}

int mdio_parse_reg(const char *str, uint16_t *reg, bool is_c45)
{
	unsigned long r;
	char *end;

	r = strtoul(str, &end, 0);
	if (end[0]) {
		fprintf(stderr, "ERROR: \"%s\" is not a valid register", str);
		return EINVAL;
	}

	if (r > (is_c45 ? UINT16_MAX : 31)) {
		fprintf(stderr, "ERROR: Register %lu is out of range [0-%u]",
			r, is_c45 ? UINT16_MAX : 31);
		return ERANGE;
	}

	*reg = r;
	return 0;
}

int mdio_parse_reg_range(const char *str, uint16_t *regs, uint16_t *rege,
			 bool is_c45)
{
	unsigned long s, e = 0;
	char *delim, *end;

	s = strtoul(str, &delim, 0);
	switch (*delim) {
	case '\0':
		e = is_c45 ? 31 : s + 127;
		break;
	case '+':
		e += s;
		/* fall-through */
	case '-':
		e += strtoul(delim + 1, &end, 0);

		if (!(*end))
			break;

		/* fall-through */
	default:
		fprintf(stderr, "ERROR: \"%s\" is not a valid register range",
			str);
		return EINVAL;
	}


	if (e > (is_c45 ? UINT16_MAX : 31)) {
		fprintf(stderr, "ERROR: Register %lu is out of range [0-%u]",
			e, is_c45 ? UINT16_MAX : 31);
		return ERANGE;
	}

	*regs = s;
	*rege = e;
	return 0;
}

int mdio_parse_val(const char *str, uint16_t *val, uint16_t *mask)
{
	unsigned long v, m = 0;
	char *end;

	v = strtoul(str, &end, 0);
	if (!end[0])
		goto done;

	if (end[0] != '/')
		goto err_invalid;

	if (!mask) {
		fprintf(stderr, "ERROR: Masking of value not allowed");
		return EINVAL;
	}

	m = strtoul(end + 1, &end, 0);
	if (end[0])
		goto err_invalid;

done:
	if (v > UINT16_MAX) {
		fprintf(stderr, "ERROR: Value %#lx is out of range [0-%u]",
			v, UINT16_MAX);
		return ERANGE;
	}

	if (m > UINT16_MAX) {
		fprintf(stderr, "ERROR: Mask %#lx is out of range [0-%u]",
			m, UINT16_MAX);
		return ERANGE;
	}

	*val = v;
	if (mask)
		*mask = m;
	return 0;

err_invalid:
	fprintf(stderr, "ERROR: \"%s\" is not a valid register value", str);
	return EINVAL;
}

static char buf[0x1000] __attribute__ ((aligned (NLMSG_ALIGNTO)));
static const size_t len = 0x1000;
static uint16_t mdio_family;

static int parse_attrs(const struct nlattr *attr, void *data)
{
	const struct nlattr **tb = data;
	int type = mnl_attr_get_type(attr);

	tb[type] = attr;

	return MNL_CB_OK;
}

static struct mnl_socket *msg_send(int bus, struct nlmsghdr *nlh)
{
	struct mnl_socket *nl;
	int ret;

	nl = mnl_socket_open(bus);
	if (nl == NULL) {
		perror("mnl_socket_open");
		return NULL;
	}

	ret = mnl_socket_bind(nl, 0, MNL_SOCKET_AUTOPID);
	if (ret < 0) {
		perror("mnl_socket_bind");
		return NULL;
	}

	ret = mnl_socket_sendto(nl, nlh, nlh->nlmsg_len);
	if (ret < 0) {
		perror("mnl_socket_send");
		return NULL;
	}

	return nl;
}

static int msg_recv(struct mnl_socket *nl, mnl_cb_t callback, void *data, int seq)
{
	unsigned int portid;
	int ret;

	portid = mnl_socket_get_portid(nl);

	ret = mnl_socket_recvfrom(nl, buf, len);
	while (ret > 0) {
		ret = mnl_cb_run(buf, ret, seq, portid, callback, data);
		if (ret <= 0)
			break;
		ret = mnl_socket_recvfrom(nl, buf, len);
	}

	return ret;
}

static int msg_query(struct nlmsghdr *nlh, mnl_cb_t callback, void *data)
{
	unsigned int seq;
	struct mnl_socket *nl;
	int ret;

	seq = time(NULL);
	nlh->nlmsg_seq = seq;

	nl = msg_send(NETLINK_GENERIC, nlh);
	if (!nl)
		return -ENOTSUP;

	ret = msg_recv(nl, callback, data, seq);
	mnl_socket_close(nl);
	return ret;
}


struct nlmsghdr *msg_init(int cmd, int flags)
{
	struct genlmsghdr *genl;
	struct nlmsghdr *nlh;

	nlh = mnl_nlmsg_put_header(buf);
	if (!nlh)
		return NULL;

	nlh->nlmsg_type	 = mdio_family;
	nlh->nlmsg_flags = flags;

	genl = mnl_nlmsg_put_extra_header(nlh, sizeof(struct genlmsghdr));
	genl->cmd = cmd;
	genl->version = 1;

	return nlh;
}

void mdio_prog_push(struct mdio_prog *prog, struct mdio_nl_insn insn)
{
	prog->insns = realloc(prog->insns, (++prog->len) * sizeof(insn));
	memcpy(&prog->insns[prog->len - 1], &insn, sizeof(insn));
}

int mdio_raw_read_cb(uint32_t *data, int len, int err, void *_null)
{
	if (len != 1)
		return 1;

	printf("0x%4.4x\n", *data);
	return err;
}

int mdio_raw_write_cb(uint32_t *data, int len, int err, void *_null)
{
	if (len != 0)
		return 1;

	return err;
}

int mdio_raw_exec(struct mdio_ops *ops, int argc, char **argv)
{
	struct mdio_prog prog = MDIO_PROG_EMPTY;
	mdio_xfer_cb_t cb = mdio_raw_read_cb;
	uint16_t dev, reg, val, mask;
	int err;

	switch (argc) {
	case 3:
		cb = mdio_raw_write_cb;
		if (mdio_parse_val(argv[2], &val, &mask))
			return 1;

		/* fall-through */
	case 2:
		if (mdio_parse_dev(argv[0], &dev, true))
			return 1;

		if (mdio_parse_reg(argv[1], &reg, dev & MDIO_PHY_ID_C45))
			return 1;

		break;

	default:
		ops->usage(stderr);
		return 1;
	}

	if ((cb == mdio_raw_write_cb) && mask) {
		err = ops->push_read(ops, &prog, dev, reg);
		if (err)
			return err;
		mdio_prog_push(&prog, INSN(AND, REG(0), IMM(mask), REG(0)));
		mdio_prog_push(&prog, INSN(OR,  REG(0), IMM(val),  REG(0)));
		err = ops->push_write(ops, &prog, dev, reg, REG(0));
		if (err)
			return err;
	} else if (cb == mdio_raw_write_cb) {
		err = ops->push_write(ops, &prog, dev, reg, IMM(val));
		if (err)
			return err;
	} else {
		err = ops->push_read(ops, &prog, dev, reg);
		if (err)
			return err;
		mdio_prog_push(&prog, INSN(EMIT,  REG(0),   0,         0));
	}

	err = mdio_xfer(ops->bus, &prog, cb, NULL);
	free(prog.insns);
	if (err) {
		fprintf(stderr, "ERROR: Raw operation failed (%d)\n", err);
		return 1;
	}

	return 0;
}

int mdio_device_dflt_parse_reg(struct mdio_device *dev,
			       int *argcp, char ***argvp,
			       uint32_t *regs, uint32_t *rege)
{
	unsigned long long r;
	char *str, *end;

	if (rege) {
		fprintf(stderr, "ERROR: Implement ranges\n");
		return ENOSYS;
	}

	str = argv_pop(argcp, argvp);
	if (!str) {
		fprintf(stderr, "ERROR: Expected register\n");
		return EINVAL;
	}

	r = strtoull(str, &end, 0);
	if (end[0]) {
		fprintf(stderr, "ERROR: \"%s\" is not a valid register\n", str);
		return EINVAL;
	}

	if (r > dev->mem.max) {
		fprintf(stderr, "ERROR: Register %llu is out of range "
			"[0-%"PRIu32"]\n", r, dev->mem.max);
		return ERANGE;
	}

	*regs = r;
	return 0;
}

int mdio_device_parse_reg(struct mdio_device *dev, int *argcp, char ***argvp,
			  uint32_t *regs, uint32_t *rege)
{
	if (dev->driver->parse_reg)
		return dev->driver->parse_reg(dev, argcp, argvp, regs, rege);

	return mdio_device_dflt_parse_reg(dev, argcp, argvp, regs, rege);
}

int mdio_device_dflt_parse_val(struct mdio_device *dev,
			       int *argcp, char ***argvp,
			       uint32_t *val, uint32_t *mask)
{
	unsigned long long max = (1 << dev->mem.width) - 1;
	unsigned long long v, m = 0;
	char *str, *end;

	if (dev->driver->parse_val)
		return dev->driver->parse_val(dev, argcp, argvp, val, mask);

	str = argv_pop(argcp, argvp);
	if (!str) {
		fprintf(stderr, "ERROR: Expected register\n");
		return EINVAL;
	}

	v = strtoull(str, &end, 0);
	if (!end[0])
		goto done;

	if (end[0] != '/')
		goto err_invalid;

	if (!mask) {
		fprintf(stderr, "ERROR: Masking of value not allowed");
		return EINVAL;
	}

	m = strtoull(end + 1, &end, 0);
	if (end[0])
		goto err_invalid;

done:
	if (v > max) {
		fprintf(stderr, "ERROR: Value %#llx is out of range "
			"[0-%#llx]", v, max);
		return ERANGE;
	}

	if (m > max) {
		fprintf(stderr, "ERROR: Mask %#llx is out of range "
			"[0-%#llx]", m, max);
		return ERANGE;
	}

	*val = v;
	if (mask)
		*mask = m;
	return 0;

err_invalid:
	fprintf(stderr, "ERROR: \"%s\" is not a valid register value\n", str);
	return EINVAL;
}

int mdio_device_parse_val(struct mdio_device *dev, int *argcp, char ***argvp,
			  uint32_t *val, uint32_t *mask)
{
	if (dev->driver->parse_val)
		return dev->driver->parse_val(dev, argcp, argvp, val, mask);

	return mdio_device_dflt_parse_val(dev, argcp, argvp, val, mask);
}

int mdio_common_raw_exec(struct mdio_device *dev, int argc, char **argv)
{
	struct mdio_prog prog = MDIO_PROG_EMPTY;
	mdio_xfer_cb_t cb = mdio_raw_read_cb;
	uint32_t reg, val, mask;
	int err;

	err = mdio_device_parse_reg(dev, &argc, &argv, &reg, NULL);
	if (err)
		return err;

	if (argv_peek(argc, argv)) {
		cb = mdio_raw_write_cb;

		err = mdio_device_parse_val(dev, &argc, &argv, &val, &mask);
		if (err)
			return err;
	}

	if (argv_peek(argc, argv)) {
		fprintf(stderr, "ERROR: Unexpected argument");
		return EINVAL;
	}

	if ((cb == mdio_raw_write_cb) && mask) {
		err = dev->driver->read(dev, &prog, reg);
		if (err)
			return err;
		mdio_prog_push(&prog, INSN(AND, REG(0), IMM(mask), REG(0)));
		mdio_prog_push(&prog, INSN(OR,  REG(0), IMM(val),  REG(0)));
		err = dev->driver->write(dev, &prog, reg, REG(0));
		if (err)
			return err;
	} else if (cb == mdio_raw_write_cb) {
		err = dev->driver->write(dev, &prog, reg, IMM(val));
		if (err)
			return err;
	} else {
		err = dev->driver->read(dev, &prog, reg);
		if (err)
			return err;
		mdio_prog_push(&prog, INSN(EMIT,  REG(0),   0,         0));
	}

	err = mdio_xfer(dev->bus, &prog, cb, NULL);
	free(prog.insns);
	if (err) {
		fprintf(stderr, "ERROR: Raw operation failed (%d)\n", err);
		return 1;
	}

	return 0;
}

int mdio_common_bench_cb(uint32_t *data, int len, int err, void *_start)
{
	struct timespec end, *start = _start;

	clock_gettime(CLOCK_MONOTONIC, &end);

	if (len) {
		int i;

		printf("Read back %d incorrect values:\n", len);
		err = 1;

		for (i = 0; i < len; i++)
			printf("\t0x%4.4x\n", data[i]);
	}

	end.tv_sec -= start->tv_sec;
	end.tv_nsec -= start->tv_nsec;
	if (end.tv_nsec < 0) {
		end.tv_nsec += 1000000000;
		end.tv_sec--;
	}

	if (err)
		printf("Benchmark failed after ");
	else
		printf("Performed 1000 reads in ");

	if (end.tv_sec)
		printf("%"PRId64".%2.2"PRId64"s\n", (int64_t)end.tv_sec, (int64_t)end.tv_nsec / 10000000);
	else if (end.tv_nsec > 1000000)
		printf("%"PRId64"ms\n", (int64_t)end.tv_nsec / 1000000);
	else if (end.tv_nsec > 1000)
		printf("%"PRId64"us\n", (int64_t)end.tv_nsec / 1000);
	else
		printf("%"PRId64"ns\n", (int64_t)end.tv_nsec);

	return err;
}

int mdio_common_bench_exec(struct mdio_device *dev, int argc, char **argv)
{
	struct mdio_prog prog = MDIO_PROG_EMPTY;
	struct timespec start;
	uint32_t reg, val = 0;
	int err, loop;

	err = mdio_device_parse_reg(dev, &argc, &argv, &reg, NULL);
	if (err)
		return err;

	if (argv_peek(argc, argv)) {
		err = mdio_device_parse_val(dev, &argc, &argv, &val, NULL);
		if (err)
			return err;

		mdio_prog_push(&prog, INSN(ADD, IMM(val), IMM(0), REG(7)));
		err = dev->driver->write(dev, &prog, reg, REG(7));
		if (err)
			return err;
	} else {
		err = dev->driver->read(dev, &prog, reg);
		if (err)
			return err;

		mdio_prog_push(&prog, INSN(ADD, REG(0), IMM(0), REG(7)));
	}

	if (argv_peek(argc, argv)) {
		fprintf(stderr, "ERROR: Unexpected argument");
		return EINVAL;
	}


	mdio_prog_push(&prog, INSN(ADD, IMM(0), IMM(0), REG(6)));

	loop = prog.len;

	err = dev->driver->read(dev, &prog, reg);
	if (err)
		return err;

	mdio_prog_push(&prog, INSN(JEQ, REG(0), REG(7), IMM(1)));
	mdio_prog_push(&prog, INSN(EMIT, REG(0), 0, 0));
	mdio_prog_push(&prog, INSN(ADD, REG(6), IMM(1), REG(6)));
	mdio_prog_push(&prog, INSN(JNE, REG(6), IMM(1000), GOTO(prog.len, loop)));

	clock_gettime(CLOCK_MONOTONIC, &start);
	err = mdio_xfer_timeout(dev->bus, &prog, mdio_common_bench_cb, &start, 10000);
	free(prog.insns);
	if (err) {
		fprintf(stderr, "ERROR: Bench operation failed (%d)\n", err);
		return 1;
	}

	return 0;
}

int mdio_common_exec(struct mdio_device *dev, int argc, char **argv)
{
	if (!argc)
		return 1;

	if (!strcmp(argv[0], "raw")) {
		argv_pop(&argc, &argv);
		return mdio_common_raw_exec(dev, argc, argv);
	} else if (!strcmp(argv[0], "bench")) {
		argv_pop(&argc, &argv);
		return mdio_common_bench_exec(dev, argc, argv);
	}

	return mdio_common_raw_exec(dev, argc, argv);
}



struct mdio_xfer_data {
	mdio_xfer_cb_t cb;
	void *arg;
	int err;
};

static int mdio_xfer_cb(const struct nlmsghdr *nlh, void *_xfer)
{
	struct genlmsghdr *genl = mnl_nlmsg_get_payload(nlh);
	struct nlattr *tb[MDIO_NLA_MAX + 1] = {};
	struct mdio_xfer_data *xfer = _xfer;
	uint32_t *data;
	int len, err;

	mnl_attr_parse(nlh, sizeof(*genl), parse_attrs, tb);

	if (tb[MDIO_NLA_ERROR])
		xfer->err = (int)mnl_attr_get_u32(tb[MDIO_NLA_ERROR]);

	if (!tb[MDIO_NLA_DATA])
		return MNL_CB_ERROR;

	len = mnl_attr_get_payload_len(tb[MDIO_NLA_DATA]) / sizeof(uint32_t);
	data = mnl_attr_get_payload(tb[MDIO_NLA_DATA]);

	err = xfer->cb(data, len, xfer->err, xfer->arg);
	return err ? MNL_CB_ERROR : MNL_CB_OK;
}

int mdio_xfer_timeout(const char *bus, struct mdio_prog *prog,
		      mdio_xfer_cb_t cb, void *arg, uint16_t timeout_ms)
{
	struct mdio_xfer_data xfer = { .cb = cb, .arg = arg };
	struct nlmsghdr *nlh;
	int err;

	nlh = msg_init(MDIO_GENL_XFER, NLM_F_REQUEST | NLM_F_ACK);
	if (!nlh)
		return -ENOMEM;

	mnl_attr_put_strz(nlh, MDIO_NLA_BUS_ID, bus);
	mnl_attr_put(nlh, MDIO_NLA_PROG, prog->len * sizeof(*prog->insns),
		     prog->insns);

	mnl_attr_put_u16(nlh, MDIO_NLA_TIMEOUT, timeout_ms);

	err = msg_query(nlh, mdio_xfer_cb, &xfer);
	return xfer.err ? : err;
}

int mdio_xfer(const char *bus, struct mdio_prog *prog,
	      mdio_xfer_cb_t cb, void *arg)
{
	return mdio_xfer_timeout(bus, prog, cb, arg, 1000);
}

int mdio_for_each(const char *match,
		  int (*cb)(const char *bus, void *arg), void *arg)
{
	char gmatch[0x80];
	glob_t gl;
	size_t i;
	int err;

	snprintf(gmatch, sizeof(gmatch), "/sys/class/mdio_bus/%s", match);
	glob(gmatch, 0, NULL, &gl);

	for (err = 0, i = 0; i < gl.gl_pathc; i++) {
		err = cb(&gl.gl_pathv[i][strlen("/sys/class/mdio_bus/")], arg);
		if (err)
			break;
	}

	globfree(&gl);
	return err;
}


static int family_id_cb(const struct nlmsghdr *nlh, void *_null)
{
	struct genlmsghdr *genl = mnl_nlmsg_get_payload(nlh);
	struct nlattr *tb[CTRL_ATTR_MAX + 1] = {};

	mnl_attr_parse(nlh, sizeof(*genl), parse_attrs, tb);
	if (!tb[CTRL_ATTR_FAMILY_ID])
		return MNL_CB_ERROR;

	mdio_family = mnl_attr_get_u16(tb[CTRL_ATTR_FAMILY_ID]);
	return MNL_CB_OK;
}

int mdio_modprobe(void)
{
	int wstatus;
	pid_t pid;

	pid = fork();
	if (pid < 0) {
		return -errno;
	} else if (!pid) {
		execl("/sbin/modprobe", "modprobe", "mdio-netlink", NULL);
		exit(1);
	}

	if (waitpid(pid, &wstatus, 0) <= 0)
		return -ECHILD;

	if (WIFEXITED(wstatus) && !WEXITSTATUS(wstatus))
		return 0;

	return -EPERM;
}

int mdio_init(void)
{
	struct genlmsghdr *genl;
	struct nlmsghdr *nlh;

	nlh = mnl_nlmsg_put_header(buf);
	nlh->nlmsg_type	= GENL_ID_CTRL;
	nlh->nlmsg_flags = NLM_F_REQUEST | NLM_F_ACK;

	genl = mnl_nlmsg_put_extra_header(nlh, sizeof(struct genlmsghdr));
	genl->cmd = CTRL_CMD_GETFAMILY;
	genl->version = 1;

	mnl_attr_put_u16(nlh, CTRL_ATTR_FAMILY_ID, GENL_ID_CTRL);
	mnl_attr_put_strz(nlh, CTRL_ATTR_FAMILY_NAME, "mdio");

	return msg_query(nlh, family_id_cb, NULL);
}
